test/unit/AssetManagedPair.t.sol

```diff
@@ -0,0 +1,194 @@
+pragma solidity 0.8.13;
+
+import "test/__fixtures/BaseTest.sol";
+
+import { IERC20 } from "@openzeppelin/token/ERC20/IERC20.sol";
+
+import { IAssetManagedPair } from "src/interfaces/IAssetManagedPair.sol";
+import { AssetManager } from "test/__mocks/AssetManager.sol";
+
+contract AssetManagedPairTest is BaseTest
+{
+    AssetManager private _manager = new AssetManager();
+
+    IAssetManagedPair[] internal _pairs;
+
+    function setUp() public
+    {
+        _pairs.push(_constantProductPair);
+        _pairs.push(_stablePair);
+    }
+
+    function _setManager(IAssetManagedPair aPair) internal
+    {
+        // sanity
+        assertEq(address(aPair.assetManager()), address(0));
+
+        // act
+        vm.prank(address(_factory));
+        aPair.setManager(_manager);
+
+        // assert
+        assertEq(address(aPair.assetManager()), address(_manager));
+    }
+
+    function testSetManager() external
+    {
+        for (uint i = 0; i < _pairs.length; ++i) {
+            _setManager(_pairs[i]);
+        }
+    }
```

Comment:
Try the modifier approach until foundry has table tests implemented:

https://github.com/foundry-rs/foundry/issues/858#issuecomment-1236870228

Reply:
okay I tried it, it works, but it doesn't work if we require a clean slate for the test.

In other words, the state persists in between the different pairs. 

This is the same problem as the approach I had previously actually. I ran tests to see for myself that both approaches have the same problem

Reply:
Damn, that sucks. Well until there a table tests in foundry we have two options. **1.** duplicate the test cases per pair in each pairs test file. or **2**, create an asset management test file that uses this approach to test both pairs:
```
function testSetManager_CaseA()
{
    _testSetManager_CaseA(_constantPair);
    _testSetManager_CaseA(_stablePair);
}

function _testSetManager_CaseA(IAssetManagedPair aPair) private
{
    // run test
}
```

Let me know which approach you're leaning towards, I lean towards option 2

Reply:
okay I tried the second approach, it's the same problem as using the modifier. Cuz at the end of the day, forge creates a fresh instance for each public `test*` function there is. 

So I think it's only option 1 for now

Reply:
You can use `vm.revertTo` to deal with this (could even use the modifier approach that snapshots, runs test, reverts:

```solidity
// SPDX-License-Identifier: UNLICENSED
pragma solidity ^0.8.15;

import "test/__helpers/BaseTest.sol";
import {MintableERC20} from "test/__mocks/MintableERC20.sol";

import {IUniswapV2Factory} from "uni-v2-solc8/interfaces/IUniswapV2Factory.sol";
import {IFindFee} from "src/find-fee/IFindFee.sol";
import {FindFee} from "src/find-fee/FindFee.sol";

struct TestCase {
    string network;
    address target;
    bool isFactory;
    uint256 fee;
}

struct Fork {
    bool created;
    uint256 forkId;
}

contract FindFeeTest is BaseTest {
    IFindFee private _findFee;
    TestCase[] private _testCases;

    mapping(string => Fork) private _forks;

    function setUp() external {
        _testCases.push(
            TestCase({network: "ethereum", target: 0x111111111117dC0aa78b770fA6A738034120C302, isFactory: false, fee: 0})
        );
        _testCases.push(
            TestCase({network: "ethereum", target: 0x5C69bEe701ef814a2B6a3EDD4B1652CB9cc5aA6f, isFactory: true, fee: 30})
        );
        _testCases.push(
            TestCase({network: "ethereum", target: 0xC0AEe478e3658e2610c5F7A4A2E1777cE9e4f2Ac, isFactory: true, fee: 30})
        );
        _testCases.push(
            TestCase({network: "ethereum", target: 0x115934131916C8b277DD010Ee02de363c09d037c, isFactory: true, fee: 30})
        );
        _testCases.push(
            TestCase({network: "ethereum", target: 0x9DEB29c9a4c7A88a3C0257393b7f3335338D9A9D, isFactory: true, fee: 30})
        );
        _testCases.push(
            TestCase({network: "ethereum", target: 0x91fAe1bc94A9793708fbc66aDcb59087C46dEe10, isFactory: true, fee: 10})
        );
        _testCases.push(
            TestCase({network: "moonriver", target: 0x28Eaa01DC747C4e9D37c5ca473E7d167E90F8d38, isFactory: true, fee: 30})
        );
    }

    function _setUpRPC(TestCase memory aCase) private {
        // todo: we should also snapshot before each case
        Fork memory lFork = _forks[aCase.network];

        if (lFork.created == false) {
            uint256 lForkId = vm.createFork(vm.rpcUrl(aCase.network));

            lFork = Fork({created: true, forkId: lForkId});

            _forks[aCase.network] = lFork;
            console2.log("fork created for network:", aCase.network);
        }

        vm.selectFork(lFork.forkId);
    }

    function _runCase(TestCase memory aCase) private {
        if (aCase.isFactory == true) {
            (bool lSuccess, uint256 lFee) = _findFee.findFee(IUniswapV2Factory(aCase.target));

            assertEq(lSuccess, aCase.isFactory);
            assertEq(lFee, aCase.fee);
        } else {
            vm.expectRevert();
            _findFee.findFee(IUniswapV2Factory(aCase.target));
        }
    }

    function testFindFee_AllNetworks() external {
        for (uint256 i = 0; i < _testCases.length; ++i) {
            TestCase memory lCase = _testCases[i];

            _setUpRPC(lCase);

            uint256 lBefore = vm.snapshot();
            _findFee = new FindFee();

            _runCase(lCase);

            vm.revertTo(lBefore);
        }
    }
}
````

Reply:
alright very nice. I implemented a POC that combines the `snapshot` and `revertTo` approach with the modifier approach you shared previously. 

I can't do this for all the tests yet due to the interface problem. `IAssetManagedPair` doesn't have `mint` and `burn` so I can't convert all of them yet. Once we get the interfaces straight I'll convert all applicable tests into this form

Reply:
Sounds good, will you delay this merge or merge this and then do a subsequent PR? If doing the subsequent PR, just make sure its tracked either by TODO comment, issue, or personal notes

Reply:
yes I will do it in the same PR as the class + interface refactoring. Captured in #76